File: src/Scripting/PHP/adapter.php
<?php

declare(strict_types=1);

interface TemperatureReader
{
    public function readCelsius(): int;
}

final class LegacyFahrenheitSensor
{
    public function readFahrenheit(): int
    {
        return 86;
    }
}

final class FahrenheitSensorAdapter implements TemperatureReader
{
    public function __construct(private readonly LegacyFahrenheitSensor $sensor)
    {
    }

    public function readCelsius(): int
    {
        return intdiv(($this->sensor->readFahrenheit() - 32) * 5, 9);
    }
}

$legacy = new LegacyFahrenheitSensor();
$reader = new FahrenheitSensorAdapter($legacy);
printf("legacy=%dF\n", $legacy->readFahrenheit());
printf("adapted=%dC\n", $reader->readCelsius());

File: src/Scripting/Lua/adapter.lua
local legacy = {
  read_fahrenheit = function()
    return 86
  end
}

local function adapt_temperature(sensor)
  return {
    read_celsius = function()
      return math.floor(((sensor.read_fahrenheit() - 32) * 5 / 9) + 0.5)
    end
  }
end

local reader = adapt_temperature(legacy)
print(string.format("legacy=%dF", legacy.read_fahrenheit()))
print(string.format("adapted=%dC", reader.read_celsius()))
